### PR TITLE
fix: default not shrink for custom canisters

### PR DESCRIPTION
**Pull request by lwshang**
# Description

For custom canisters, `dfx` should better not do shrink by default. Developer still can enable it in `dfx.json`.

# How Has This Been Tested?

An e2e test was added.

# Checklist:

- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [x] I have edited the CHANGELOG accordingly.
- [x] I have made corresponding changes to the documentation.

### CODE REVIEW

File: src/dfx/src/lib/canister_info.rs [hunk 1]
```diff
@@ -202,7 +202,7 @@ impl CanisterInfo {
         &self.args
     }
 
-    pub fn get_shrink(&self) -> bool {
+    pub fn get_shrink(&self) -> Option<bool> {
         self.shrink
     }
```
Comment (ericswanson-dfinity) [inline src/dfx/src/lib/canister_info.rs:207]: I'm having a hard time following `!matches!(canister_info.get_shrink(), Some(false))`.  How about this?
```suggestion
    pub fn get_shrink(&self) -> bool {
        match self.type_specific {
            CanisterTypeProperties::Rust { .. } | CanisterTypeProperties::Motoko => {
                self.shrink.unwrap_or(true)
            }
            CanisterTypeProperties::Custom { .. } | CanisterTypeProperties::Assets { .. } => {
                self.shrink.unwrap_or(false)
            }
        }
    }
```

Comment (adamspofford-dfinity) [reply src/dfx/src/lib/canister_info.rs:207]: For future reference, `== Some(true)` works fine too.

Comment (lwshang) [reply src/dfx/src/lib/canister_info.rs:207]: That `matches!` actually came from `cargo clippy`.

While moving such canister type specific logic into the general `CanisterInfo` is not good for encapsulation.
I would still keep the logic in each canister type builder and replace `matches` with more readable syntax.

File: src/dfx/src/lib/builders/motoko.rs [hunk 2]
```diff
@@ -166,7 +166,9 @@ impl CanisterBuilder for MotokoBuilder {
         };
         motoko_compile(&self.logger, cache.as_ref(), &params)?;
 
-        if canister_info.get_shrink() {
+        // Default to shrink Motoko canister unless explicitly disable
+        let shrink = canister_info.get_shrink() != Some(false);
```
Comment (ericswanson-dfinity) [inline src/dfx/src/lib/builders/motoko.rs:170]: This is better, but still hard to read.  Using `unwrap_or` removes the need for the comment, even.
```suggestion
        let shrink = canister_info.get_shrink().unwrap_or(true);
```

File: src/dfx/src/lib/builders/custom.rs [hunk 3]
```diff
@@ -140,8 +140,10 @@ impl CanisterBuilder for CustomBuilder {
             }
         }
 
+        // Default **NOT** to shrink custom canister unless explicitly enable
+        let shrink = info.get_shrink() == Some(true);
```
Comment (ericswanson-dfinity) [inline src/dfx/src/lib/builders/custom.rs:144]:
```suggestion
        let shrink = info.get_shrink().unwrap_or(false);
```

File: src/dfx/src/lib/builders/rust.rs [hunk 4]
```diff
@@ -92,7 +92,9 @@ impl CanisterBuilder for RustBuilder {
         );
         let output = cargo.output().context("Failed to run 'cargo build'. You might need to run `cargo update` (or a similar command like `cargo vendor`) if you have updated `Cargo.toml`, because `dfx build` uses the --locked flag with Cargo.")?;
 
-        if canister_info.get_shrink() {
+        // Default to shrink Rust canister unless explicitly disable
+        let shrink = canister_info.get_shrink() != Some(false);
+        if shrink {
```
Comment (ericswanson-dfinity) [inline src/dfx/src/lib/builders/rust.rs:97]:
```suggestion
        let shrink = canister_info.get_shrink().unwrap_or(true);
        if shrink {
```

**Review by ericswanson-dfinity — CHANGES_REQUESTED** [review]

**Review by ericswanson-dfinity — APPROVED** [review]

**Pull request merged**